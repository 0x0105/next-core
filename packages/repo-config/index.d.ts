export const getEasyopsConfig: () => {
  usePublicScope: boolean;
  useLocalSdk: boolean;
  contractUrl: string;
  contractYamlDir: string;
};
