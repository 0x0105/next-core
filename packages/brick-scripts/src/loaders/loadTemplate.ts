import fs from "fs-extra";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";
import rp from "request-promise-native";
import { FileWithContent, TargetType } from "../interface";
import { getEasyopsConfig } from "../getEasyopsConfig";
const { exec } = require("child_process");
// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const easyopsConfig = getEasyopsConfig();
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { templateDependencies } = packageJson;
const basicBricksVersion = templateDependencies["@bricks/basic-bricks"];
const brickContainerVersion = templateDependencies["@easyops/brick-container"];

const workspacePackageJson = JSON.parse(
  fs.readFileSync(path.join(process.cwd(), "package.json"), "utf8")
);
const workspaceHomepage = workspacePackageJson.homepage;
let userName = "bot";
exec("git config user.name", { encoding: "utf8" }, function (err, value) {
  if (err) return;
  userName = value.trim();
});

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter((key) =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string): string {
  const pkg = JSON.parse(jsonString);

  const peerDeps = pkg.peerDependencies || {};
  if (peerDeps["@bricks/basic-bricks"]) {
    peerDeps["@bricks/basic-bricks"] = basicBricksVersion;
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export async function loadTemplate({
  targetType,
  packageName,
  brickName,
  templateName,
  processorName,
  targetRoot,
  docRoot,
}: {
  targetType: TargetType;
  packageName: string;
  brickName: string;
  templateName: string;
  processorName: string;
  targetRoot: string;
  docRoot: string;
}): Promise<FileWithContent[]> {
  const targetMap: { [key: string]: string } = {
    [TargetType.A_NEW_BRICK]: "brick",
    [TargetType.A_NEW_CUSTOM_TEMPLATE]: "custom-template",
    [TargetType.A_NEW_PACKAGE_OF_BRICKS]: "bricks-pkg",
    [TargetType.A_NEW_PACKAGE_OF_LIBS]: "libs-pkg",
    [TargetType.A_NEW_PACKAGE_OF_MICRO_APPS]: "micro-apps-pkg",
    [TargetType.A_NEW_CUSTOM_PROVIDER_BRICK]: "custom-provider-brick",
    [TargetType.A_NEW_CUSTOM_PROCESSOR]: "custom-processor",
    [TargetType.A_NEW_PACKAGE_OF_PROVIDERS]: "providers-pkg",
    [TargetType.A_NEW_PACKAGE_OF_DLL]: "dll-pkg",
    [TargetType.TRANSFORM_A_MICRO_APP]: "transformed-micro-apps-pkg",
    [TargetType.A_NEW_LEGACY_TEMPLATE]: "template",
    [TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES]: "templates-pkg",
    [TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES]:
      "i18n-patched-templates-pkg",
  };
  const templateDirOfFileName = targetMap[targetType];
  const templateRoot = path.join(__dirname, "../../template");
  let templateDir = path.join(templateRoot, templateDirOfFileName);

  const ignores = [".DS_Store"];
  let sdkName: string;
  if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库直接使用构件库的模板文件，但忽略其 `src` 目录。
    templateDir = path.join(
      templateRoot,
      targetMap[TargetType.A_NEW_PACKAGE_OF_BRICKS]
    );
    ignores.push("template/bricks-pkg/src/");
    sdkName = `${packageName.replace(/^providers-of-/, "")}-sdk`;
  }

  const translations: Record<string, string> = {
    "$workspace-homepage$": workspaceHomepage,
    "$kebab-package-name$": packageName,
    $PascalPackageName$: changeCase.pascalCase(packageName),
    "$Title Package Name$": changeCase.capitalCase(packageName),
    $CONSTANT_PACKAGE_NAME$: changeCase.constantCase(packageName),
    $camelCasePackageName$: changeCase.camelCase(packageName),
    $PascalBrickName$: changeCase.pascalCase(brickName),
    "$kebab-brick-name$": `${packageName}.${brickName}`,
    "$kebab-brick-last-name$": brickName,
    "$kebab-custom-provider-brick-name$": `${packageName}.provider-${brickName}`,
    "$generator.version$": `v${packageJson.version}`,
    "$brick.container.version$": brickContainerVersion,
    "$kebab-sdk-name$": sdkName,
    "$kebab-template-name$": templateName,
    "$kebab-username$": userName,
    $camelTemplateName$: changeCase.camelCase(templateName),
    $PascalTemplateName$: changeCase.pascalCase(templateName),
    $camelProcessorName$: changeCase.camelCase(processorName),
  };

  const filter = (src: string): boolean =>
    ignores.every((item) => !src.includes(item));

  const templateGroups: any = [
    {
      templateDir,
      targetDir: targetRoot,
      files: klawSync(templateDir, {
        depthLimit: 4,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    },
  ];

  if (targetType === TargetType.A_NEW_PACKAGE_OF_BRICKS && brickName) {
    // Also create a new brick for the new bricks-package
    const brickTemplateDir = path.join(templateRoot, "brick");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: path.join(targetRoot, "src", brickName),
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES) {
    // Also create a new brick for the new bricks-package
    const templateTemplateDir = path.join(templateRoot, "template");
    templateGroups.push({
      templateDir: templateTemplateDir,
      targetDir: path.join(targetRoot, "src"),
      files: klawSync(templateTemplateDir, {
        depthLimit: 1,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库还有专属的一些模板文件。
    const brickTemplateDir = path.join(templateRoot, "providers-pkg");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: targetRoot,
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });

    let sdkVersion;
    const getVersion = async () => {
      try {
        const sdkPackage = await rp({
          url: `https://registry.npm.easyops.local/@sdk/${sdkName}`,
          json: true,
          strictSSL: false,
        });
        sdkVersion = `^${sdkPackage["dist-tags"].latest}`;
      } catch {
        sdkVersion = "FETCH LATEST VERSION ERROR";
      }
      translations["$sdk.version$"] = sdkVersion;
    };

    if (!easyopsConfig?.useLocalSdk) {
      await getVersion();
    } else {
      translations["$sdk.version$"] = "1.0.0";
    }
  }

  const files: FileWithContent[] = templateGroups.reduce(
    (acc: FileWithContent[], group: any) =>
      acc.concat(
        group.files.map((file: any) => {
          const content = replaceFileContent(file.path, translations);

          let realPath = file.path;
          const basename = path.basename(file.path);
          const filenameFinds = Object.keys(translations).filter((key) =>
            basename.includes(key)
          );
          if (filenameFinds.length > 0) {
            realPath = path.join(
              path.dirname(realPath),
              filenameFinds.reduce(
                (acc, find) => replaceAll(acc, find, translations[find]),
                basename
              )
            );
          }

          return [
            path.join(
              group.targetDir,
              path.relative(group.templateDir, realPath)
            ),
            content,
          ];
        })
      ),
    []
  );

  if (
    targetType !== TargetType.A_NEW_BRICK &&
    targetType !== TargetType.A_NEW_CUSTOM_TEMPLATE &&
    targetType !== TargetType.A_NEW_CUSTOM_PROVIDER_BRICK &&
    targetType !== TargetType.A_NEW_CUSTOM_PROCESSOR &&
    targetType !== TargetType.A_NEW_LEGACY_TEMPLATE &&
    targetType !== TargetType.TRANSFORM_A_MICRO_APP &&
    targetType !== TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES
  ) {
    files.push([
      path.join(targetRoot, "package.json"),
      replaceDepsVersion(
        replaceFileContent(
          path.join(templateRoot, `${targetMap[targetType]}.json`),
          translations
        )
      ),
    ]);
  }

  return files;
}
