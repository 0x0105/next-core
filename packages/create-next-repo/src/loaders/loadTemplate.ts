import fs from "fs";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";

type FileWithContent = [string, string];

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const devDependenciesCopyMap = {
  devDependencies: [
    "@easyops/babel-preset-next",
    "@easyops/brick-container",
    "@easyops/brick-scripts",
    "@easyops/browserslist-config-next",
    "@easyops/build-config-factory",
    "@easyops/eslint-config-next",
    "@easyops/webpack-config-factory"
  ],
  templateDependencies: ["@bricks/basic-bricks"]
};

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter(key =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string): string {
  const pkg = JSON.parse(jsonString);
  for (const [type, deps] of Object.entries(devDependenciesCopyMap)) {
    for (const dep of deps) {
      pkg.devDependencies[dep] = packageJson[type][dep];
    }
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export function loadTemplate(
  repoName: string,
  targetDir: string,
  flags: { internal: boolean }
): FileWithContent[] {
  const templateDir = path.join(__dirname, "../../template");

  const ignores = [".DS_Store"];
  const filter = (src: string): boolean =>
    ignores.some(item => !src.includes(item));

  const templateFiles = klawSync(templateDir, {
    depthLimit: 3,
    nodir: true,
    filter: item => filter(item.path)
  });

  const translations: Record<string, string> = {
    "$kebab-repo-name$": changeCase.paramCase(repoName),
    "$Title Repo Name$": changeCase.capitalCase(repoName),
    "$generator.version$": `v${packageJson.version}`,
    "$easyops-registry$": flags.internal
      ? "http://registry.npm.easyops.local"
      : "http://r.pnpm.easyops.cn"
  };

  const files: FileWithContent[] = templateFiles.map(file => {
    const content = replaceFileContent(file.path, translations);
    return [
      path.join(targetDir, path.relative(templateDir, file.path)),
      content
    ];
  });

  files.push([
    path.join(targetDir, "package.json"),
    replaceDepsVersion(
      replaceFileContent(
        path.join(templateDir, "../template.json"),
        translations
      )
    )
  ]);

  return files;
}
