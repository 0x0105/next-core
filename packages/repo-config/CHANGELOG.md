# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.2](https://github.com/easyops-cn/next-core/compare/@next-core/repo-config@0.1.1...@next-core/repo-config@0.1.2) (2021-01-26)

**Note:** Version bump only for package @next-core/repo-config

## 0.1.1 (2021-01-22)

### Features

- rename [@easyops](https://github.com/easyops) by [@next-core](https://github.com/next-core), and [@dll](https://github.com/dll) by [@next-dll](https://github.com/next-dll) ([d87ed1e](https://github.com/easyops-cn/next-core/commit/d87ed1e))
