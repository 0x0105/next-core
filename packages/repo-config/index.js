const fs = require("fs");
const path = require("path");

exports.getEasyopsConfig = () => {
  const easyopsConfig = {
    useLocalSdk: false,
    usePublicScope: false,
    contractYamlDir: "easyops",
    contractUrl: "git@git.easyops.local:anyclouds/contract-center.git",
  };
  const configPath = path.resolve(".easyops-yo.json");
  if (fs.existsSync(configPath)) {
    const incomeConfig = JSON.parse(fs.readFileSync(configPath, "utf-8"));
    Object.assign(easyopsConfig, incomeConfig);
  }
  return easyopsConfig;
};
