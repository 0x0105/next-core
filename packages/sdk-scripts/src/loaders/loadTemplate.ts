import fs from "fs";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import changeCase from "change-case";
import { FileWithContent } from "../interface";

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { devDependencies } = packageJson;
const brickHttpVersion = devDependencies["@easyops/brick-http"];
const rollupConfigFactoryVersion =
  devDependencies["@easyops/rollup-config-factory"];

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter(key =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string, sdkVersion: string): string {
  const pkg = JSON.parse(jsonString);
  pkg.devDependencies["@easyops/brick-http"] = brickHttpVersion;
  pkg.devDependencies[
    "@easyops/rollup-config-factory"
  ] = rollupConfigFactoryVersion;
  if (sdkVersion) {
    pkg.version = sdkVersion;
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export function loadTemplate(
  serviceName: string,
  sdkRoot: string,
  sdkVersion: string
): FileWithContent[] {
  const templateDir = path.join(__dirname, "../../template");

  const ignores = [".DS_Store"];
  const filter = (src: string): boolean =>
    ignores.some(item => !src.includes(item));

  const templateFiles = klawSync(templateDir, {
    depthLimit: 2,
    nodir: true,
    filter: item => filter(item.path)
  });

  const translations: Record<string, string> = {
    "$kebab-service-name$": changeCase.kebab(serviceName),
    "$Title Service Name$": changeCase.title(serviceName),
    $PascalServiceName$: changeCase.pascal(serviceName),
    "$generator.version$": `v${packageJson.version}`
  };

  const files: FileWithContent[] = templateFiles.map(file => {
    const content = replaceFileContent(file.path, translations);
    return [path.join(sdkRoot, path.relative(templateDir, file.path)), content];
  });

  files.push([
    path.join(sdkRoot, "package.json"),
    replaceDepsVersion(
      replaceFileContent(
        path.join(templateDir, "../template.json"),
        translations
      ),
      sdkVersion
    )
  ]);

  return files;
}
