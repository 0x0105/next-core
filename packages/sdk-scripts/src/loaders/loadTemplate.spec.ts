import path from "path";
import { getEasyopsConfig } from "@next-core/repo-config";
import { loadTemplate } from "./loadTemplate";

jest.mock("@next-core/repo-config");

describe("loadTemplate", () => {
  it("should work", () => {
    (getEasyopsConfig as jest.Mock).mockReturnValue({
      usePublicScope: false,
    });
    let files = loadTemplate("cd", "/tmp/sdk", "1.2.3");
    files = files.filter(([filePath]) => {
      const fileName = path.basename(filePath);
      return !["README.md"].includes(fileName);
    });
    expect(files).toMatchSnapshot();
  });

  it("should work for public scope", () => {
    (getEasyopsConfig as jest.Mock).mockReturnValue({
      usePublicScope: true,
    });
    let files = loadTemplate("cd", "/tmp/sdk", "1.2.3");
    files = files.filter(([filePath]) => {
      const fileName = path.basename(filePath);
      return !["README.md"].includes(fileName);
    });
    expect(files).toMatchSnapshot();
  });
});
