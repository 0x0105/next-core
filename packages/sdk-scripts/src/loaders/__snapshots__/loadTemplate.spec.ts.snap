// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadTemplate should work 1`] = `
Array [
  Array [
    "/tmp/sdk/rollup.config.js",
    "import { rollupFactory } from \\"@easyops/rollup-config-factory\\";

export default rollupFactory({ umdName: \\"CdSdk\\" });
",
  ],
  Array [
    "/tmp/sdk/src/wrapper.ts",
    "export interface ResponseBodyWrapper<T> {
  code: number;
  data: T;
  error?: string;
  message?: string;
}

export interface ResponseListWrapper<T> {
  list: T[];
  page: number;
  page_size: number;
  total: number;
}
",
  ],
  Array [
    "/tmp/sdk/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [\\"./src\\", \\"../../declarations\\"]
}
",
  ],
]
`;
